### PR TITLE
fix(prettier-plugin): stop dropping decorators

### PR DESCRIPTION
**Pull request by trueadm**
## Problem

The printer had no decorator handling at all, so formatting a `.tsrx` file silently **deleted every `@decorator`** — on class declarations, methods, fields, accessors, and parameters alike. Decorators have runtime effects, so this changed what the code did.

```ts
@sealed
class A {
  @log
  method() {}
  @inject accessor x = 1;
  m(@param() a: number) {}
}
```

formatted to `class A { method() {} accessor x = 1; m(a: number) {} }` — every decorator gone, in all four positions.

`printClassDeclaration`, `printPropertyDefinition`, `printMethodDefinition`, and the parameter printers never read `node.decorators`, even though the parser populates it.

## Fix

- A `Decorator` case in the print switch, plus a single hook at the printer's one exit point that prefixes decorators onto any node carrying them — so every position is covered by construction rather than per-printer.
- Line breaking follows prettier: class decorators each take their own line; class member decorators keep the lines they were written with; an inline member decorator gets its own group, so one too long to share the member's line moves down rather than breaking apart; parameter decorators stay inline.
- Two hoists, where a decorator's owner isn't where it's printed: exported classes emit decorators above `export`, and a parameter property's decorators (which the parser hangs off the inner parameter) print before `private readonly`.
- A decorated class expression in a declarator moves below the `=` so its hardlines stay indented.
- `@tsrx/core`'s AST types gain the `Decorator` node, so the printer's handling is typed rather than cast around.

Follow-up to #1404, which fixed the same family of bugs for `readonly`/`abstract`/`declare`/`override`/`accessor` and left decorators as a separate concern.

## Divergences from vanilla prettier

I diffed 26 decorator cases against prettier's own `typescript` parser: 24 match byte-for-byte and all are idempotent. Two deliberate exceptions:

- `export @dec class A {}` normalizes to `@dec` above `export`. Both forms parse to identical ASTs here, so the position isn't recoverable; the two are equivalent TypeScript.
- `constructor(@inject private readonly x: Foo, @other b: Bar)` stays on one line where vanilla breaks it. Pre-existing and unrelated — `printMethodDefinition` joins params with a literal `', '` and never breaks them, decorators or not.

## Testing

- 18 new tests in a `describe('decorators survive formatting')` block. All 18 fail without the printer change and pass with it. The shared `format` helper asserts single-pass idempotence on every call.
- 423 prettier-plugin tests pass; `pnpm format:check` clean; `pnpm typecheck` clean repo-wide (worth running given the shared-types change).
- Full `pnpm test`: 4846 pass. The 7 failing test *files* are the pre-existing `@tsrx/eslint-parser` resolution failures in a fresh worktree (its exports point at `./dist`, which isn't built) — unrelated to this change.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> The change fixes behavior that altered emitted code (decorators), but scope is limited to the formatter with broad tests; export/parenthesis edge cases are the main review focus.
> 
> **Overview**
> **Fixes a bug where `@tsrx/prettier-plugin` stripped every decorator on format**, which could change runtime behavior because decorators are not cosmetic.
> 
> The printer now emits `Decorator` nodes and prefixes decorator lists on decorated nodes via a single exit hook, with Prettier-like breaks: one decorator per line on classes, member decorators that respect original line breaks (or wrap when too long), and inline parameter decorators. **Export** printers hoist class decorators above `export` (except parenthesized `export default`), **TS parameter properties** print inner-parameter decorators before modifiers, and **decorated class expressions** in variable declarators indent below `=`.
> 
> `@tsrx/core` gains a typed **`Decorator`** AST node for the printer. A large **`decorators survive formatting`** test suite covers round-trip and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8d7d35270d7de7ad98e532b4a04102bc5b5434f0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/prettier-plugin/src/index.js [hunk 1]
```diff
@@ -856,6 +862,134 @@ function printKey(node, path, options, print) {
...
+function decoratorsPrintedByParent(node, path) {
+	const parent = /** @type {AST.Node | null} */ (path.getParentNode());
+
+	if (!parent) {
+		return false;
+	}
+
+	if (parent.type === 'ExportNamedDeclaration' || parent.type === 'ExportDefaultDeclaration') {
+		return parent.declaration === node;
+	}
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/prettier-plugin/src/index.js:905]: I checked this against the parser and printer. The premise is correct but the conclusion doesn't hold — no change needed here. Taking the three claims in turn:

**1. Decorators are hoisted above `export default` rather than staying on the expression.**

Correct on the facts: `export default @dec class {}` really does parse to a `ClassExpression` in this parser, so `decoratorsPrintedByParent` applies. But that's a parser labelling quirk, not a semantic distinction. Per the decorators proposal, `export default @dec class Named {}` *is* a class declaration, and both TypeScript and vanilla prettier treat it as one — vanilla prints `export default\n@dec\nclass Named {}`. The output here, `@dec\nexport default class Named {}`, re-parses to a `ClassDeclaration`, which is the correct semantics. It's the same normalisation already noted in the PR description for `export @dec class A {}`, where the two forms are likewise AST-indistinguishable.

**2. It conflicts with the `VariableDeclarator` path.**

Different syntactic context — in `const A = @dec class {}` there is nothing to hoist above, so the decorators necessarily stay with the class. Both paths produce correct output, and the declarator case matches vanilla byte-for-byte:

```ts
const A =
  @dec
  class Named {};
```

**3. It can change whether the name is module-scoped after re-parse.**

This is the substantive one, and it's a real bug — but it isn't caused by this PR. The only input where the scoping actually changes is a genuinely parenthesised `export default (class Named {})`, and this printer already rewrites that with no decorator involved at all:

| input | tsrx (before this PR) | vanilla |
| --- | --- | --- |
| `export default (class Named {});` | `export default class Named {}` | `export default (class Named {});` |

Dropping those parens does change semantics — `Named` goes from being bound only inside the class body to being a module-scoped binding — but it is pre-existing and orthogonal to decorators. Making the decorator hoist more conservative would not fix it, since the rewrite happens without decorators.

So this PR neither introduces nor worsens the scoping issue. I've filed the paren drop separately rather than widening this PR's scope.
